### DISCUSSION TITLE
General improvements to the input system

### DISCUSSION THREAD

**Discussion by Dynamic-Pistol** [godotengine/godot-proposals#10914] [Engine Core]
To be direct quickly and not bore you,here are some suggestions,feel free to make proposal issues based off them

# Add composite input actions
basically,the dev should be able to setup composite actions that are made of sub actions, the following composite types are
- Axis1D (float)
- Axis2D (Vector2)
- Indexer (int)
- Combo (PackedBoolArray)
- MultiClick (bool)

## How they work:
A composite action can contain sub actions for example,
an `Axis1D` can contain positive and negative action

a composite action is called with one of the following depending on views:

- `Input.get_value("comp name", COMP_TYPE)`
- `Input.get_value_comp("Comp name)`
- `Input.get_value[Type]("compname")`

composite actions can also be obtained as input events
```gd
if event.get_value("name"):
  pass
```

sub actions of composite actions can be obtained via `"action_name:name_of_subaction"`, for example
```gd
Input.is_action_pressed("move:xPositive")
```

if a certain action has more than axis by default, it will take over the entire action/multiple axises, for example the mouse moving delta will take over the entirety of `Axis2D`

### Axis 2D
Common uses:
- Platformers
- Scrolling

Actions:
- Positive action
- Negative action

### Axis2D
Common uses:
- Top Down movement
- 3D movement

Actions:
- Positive x action
- Negative x action
- Positive y action
- Negative y action

### Indexer
Common uses:
- Hotbars
- Commands

Actions:
- custom indices amount depending on need (although default is 1-0 for inventory hotbars)

### Combo
Common uses:
- Beatem up
- Mini games

Actions:
- custom amount of buttons,but the previous one must be pressed for the next to be true

### MutliClick
Common uses:
- Editor apps

Actions:
- custom amount of buttons,but all must be pressed for it to be true

# Add Input Node

basically,instead of having each node have a `_input` function,there will be a `InputNode` that sends event signals

## Signal types
there exists different types of signals,which are called at different times,they all have `action_name`

### Began Input
Called when the input is just started

### Performed Input
Called when it the input is done

### End Input
Called when input is released

### Cancel Input
Called when input is cancelled

### Going Input
Called when input is still on going

### Value Input
called when composite action is valid

### Value Fin Input
called when composite action is done or cancelled

# Add input action interactions and modifiers

## Interactions
interactions basically change the trigger requirement for the input action, instead of me copying and pasting in different ways, check out [Unity's implementation](https://docs.unity3d.com/Packages/com.unity.inputsystem@1.11/manual/Interactions.html#predefined-interactions) and an [Unreal video](https://www.youtube.com/watch?v=MN-0otRNmZI)

## Modifiers
modifiers modify the output of the input action, same as for interactions, here is [Unity's implementation](https://docs.unity3d.com/Packages/com.unity.inputsystem@1.11/manual/Processors.html#predefined-processors) and here is an [Unreal video](https://www.youtube.com/watch?v=Pq5xu1-pltA)

## General
both interactions and modifiers should be changed at runtime, this can be useful for settings (invert cam axis)

**Comment by Abab-bk** [comment]
I think should add a resource to store inputs instead project settings.
Can add an InputContext or something, to indicate what action can be accepted, this way no need many if in _Input().
In addition, add a input node have any advantages?

**Reply by Dynamic-Pistol** [reply]
> In addition, add a input node have any advantages?

@Abab-bk you could use it for sharing nodes between npcs and players,you could for example have the input node take control of the movement system for the player and have something else take control of the npcs, this is just a simple example, also do note this is a discussion, so feel free to write your suggestions here too! i am planning on making proposals for these
